rpcclient/rawtransactions.go

```diff
@@ -880,3 +881,134 @@ func (c *Client) DecodeScriptAsync(serializedScript []byte) FutureDecodeScriptRe
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal as an array of TestMempoolAcceptResult items.
+	var results []*btcjson.TestMempoolAcceptResult
+
+	err = json.Unmarshal(response, &results)
+	if err != nil {
+		return nil, err
+	}
+
+	return results, nil
+}
+
+// TestMempoolAcceptAsync returns an instance of a type that can be used to get
+// the result of the RPC at some future time by invoking the Receive function
+// on the returned instance.
+//
+// See TestMempoolAccept for the blocking version and more details.
+func (c *Client) TestMempoolAcceptAsync(txns []*wire.MsgTx,
+	maxFeeRate float64) FutureTestMempoolAcceptResult {
+
+	// Due to differences in the testmempoolaccept API for different
+	// backends, we'll need to inspect our version and construct the
+	// appropriate request.
+	version, err := c.BackendVersion()
+	if err != nil {
+		return newFutureError(err)
+	}
+
+	log.Debugf("TestMempoolAcceptAsync: backend version %s", version)
+
+	// Exit early if the version is below 22.0.0.
+	//
+	// Based on the history of `testmempoolaccept` in bitcoind,
+	// - introduced in 0.17.0
+	// - unchanged in 0.18.0
+	// - allowhighfees(bool) param is changed to maxfeerate(numeric) in
+	//   0.19.0
+	// - unchanged in 0.20.0
+	// - added fees and vsize fields in its response in 0.21.0
+	// - allow more than one txes in param rawtx and added package-error
+	//   and wtxid fields in its response in 0.22.0
+	// - unchanged in 0.23.0
+	// - unchanged in 0.24.0
+	// - added effective-feerate and effective-includes fields in its
+	//   response in 0.25.0
+	//
+	// We decide to not support this call for versions below 22.0.0. as the
+	// request/response formats are very different.
+	if version < BitcoindPre22 {
+		err := fmt.Errorf("%w: %v", ErrBitcoindVersion, version)
+		return newFutureError(err)
+	}
+
+	// The maximum number of transactions allowed is 25.
+	if len(txns) > 25 {
+		err := fmt.Errorf("%w: too many transactions provided",
+			ErrInvalidParam)
+		return newFutureError(err)
+	}
+
+	// Exit early if an empty array of transactions is provided.
+	if len(txns) == 0 {
+		err := fmt.Errorf("%w: no transactions provided",
+			ErrInvalidParam)
+		return newFutureError(err)
+	}
+
+	// Iterate all the transactions and turn them into hex strings.
+	rawTxns := make([]string, 0, len(txns))
+	for _, tx := range txns {
+		// Serialize the transaction and convert to hex string.
+		buf := bytes.NewBuffer(make([]byte, 0, tx.SerializeSize()))
+
+		// TODO(yy): add similar checks found in `BtcDecode` to
+		// `BtcEncode` - atm it just serializes bytes without any
+		// bitcoin-specific checks.
+		if err := tx.Serialize(buf); err != nil {
+			err = fmt.Errorf("%w: %v", ErrInvalidParam, err)
+			return newFutureError(err)
+		}
+
+		rawTx := hex.EncodeToString(buf.Bytes())
+		rawTxns = append(rawTxns, rawTx)
+
+		// Sanity check the provided tx is valid, which can be removed
+		// once we have similar checks added in `BtcEncode`.
```

Comment:
wont this check be done by bitcoind once we send it the txs?

Reply:
yeah we check it here to exit early and more importantly we wanna retain the same RPC behavior for `btcd` too. Think we can also implement the checks in `btcd` to mimic the behavior there, but then it involves more code change as we need to somehow pass the `newFutureError` to the line `response, err := ReceiveFuture(r)`. So I took the easy path instead.

Reply:
makes sense! thanks 🙏